src/transaction.rs

```diff
@@ -0,0 +1,554 @@
+            StatusWords::Success => Ok(()),
+            StatusWords::AuthBlockedError => Err(Error::PinLocked),
+            StatusWords::Other(sw) if sw >> 8 == 0x63 => Err(Error::WrongPin { tries: sw & 0xf }),
+            _ => {
+                error!(
+                    "failed changing pin, token response code: {:x}.",
+                    status_words.code()
+                );
+                Err(Error::GenericError)
+            }
+        }
+    }
+
+    /// Set the management key (MGM).
+    pub fn set_mgm_key(&self, new_key: &MgmKey, touch: Option<u8>) -> Result<(), Error> {
+        let p2 = match touch.unwrap_or_default() {
+            0 => 0xff,
+            1 => 0xfe,
+            _ => {
+                return Err(Error::GenericError);
+            }
+        };
+
+        let mut data = [0u8; DES_LEN_3DES + 3];
+        data[0] = YKPIV_ALGO_3DES;
+        data[1] = YKPIV_KEY_CARDMGM;
+        data[2] = DES_LEN_3DES as u8;
+        data[3..3 + DES_LEN_3DES].copy_from_slice(new_key.as_ref());
+
+        let status_words = APDU::new(YKPIV_INS_SET_MGMKEY)
+            .params(0xff, p2)
+            .data(&data)
+            .transmit(self, 261)?
+            .status_words();
+
+        if !status_words.is_success() {
+            return Err(Error::GenericError);
+        }
+
+        Ok(())
+    }
+
+    /// Perform a YubiKey operation which requires authentication.
+    ///
+    /// This is the common backend for all public key encryption and signing
+    /// operations.
+    // TODO(tarcieri): refactor this to be less gross/coupled.
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) fn authenticated_command(
+        &self,
+        sign_in: &[u8],
+        out: &mut [u8],
+        out_len: &mut usize,
+        algorithm: u8,
+        key: u8,
+        decipher: bool,
+    ) -> Result<(), Error> {
+        let in_len = sign_in.len();
+        let mut indata = [0u8; 1024];
+        let templ = [0, YKPIV_INS_AUTHENTICATE, algorithm, key];
+        let mut len: usize = 0;
+
+        match algorithm {
+            YKPIV_ALGO_RSA1024 | YKPIV_ALGO_RSA2048 => {
+                let key_len = if algorithm == YKPIV_ALGO_RSA1024 {
+                    128
+                } else {
+                    256
+                };
+
+                if in_len != key_len {
+                    return Err(Error::SizeError);
+                }
+            }
+            YKPIV_ALGO_ECCP256 | YKPIV_ALGO_ECCP384 => {
+                let key_len = if algorithm == YKPIV_ALGO_ECCP256 {
+                    32
+                } else {
+                    48
+                };
+
+                if (!decipher && (in_len > key_len)) || (decipher && (in_len != (key_len * 2) + 1))
+                {
+                    return Err(Error::SizeError);
+                }
+            }
+            _ => return Err(Error::AlgorithmError),
+        }
+
+        let bytes = if in_len < 0x80 {
+            1
+        } else if in_len < 0xff {
```

Comment:
Hmm. Earlier I made a comment that a `< 0xff` should actually be `< 0x100`, but now I'm thinking it should be left as-is based on this code's C equivalent.